File: src/main/kotlin/io/mcm/springbootwithkotlin/specification/BookSpecification.java
package io.mcm.springbootwithkotlin.specification;

import io.mcm.springbootwithkotlin.model.BookFilter;
import io.mcm.springbootwithkotlin.model.BooksRequest;
import io.mcm.springbootwithkotlin.model.entities.Book;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.stereotype.Component;

import javax.persistence.criteria.Predicate;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

@Component
public class BookSpecification {

    public Specification<Book> getBooks(BooksRequest request) {
        return (root, query, criteriaBuilder) -> {

            if (Objects.isNull(request) || Objects.isNull(request.getBookFilter())) {
                System.out.println("The request book parameters are not provided");
                return criteriaBuilder.and(new Predicate[0]);
            }
            BookFilter bookFilter = request.getBookFilter();
            List<Predicate> predicates = new ArrayList<>();

            if (Objects.nonNull(bookFilter.getIsIssued())) {
                predicates.add(criteriaBuilder.equal(root.get("isIssued"), bookFilter.getIsIssued()));
            }
            if (Objects.nonNull(bookFilter.getAuthor()) && !bookFilter.getAuthor().isBlank()) {
                predicates.add(criteriaBuilder.like(criteriaBuilder.lower(root.get("author")),
                        "%" + bookFilter.getAuthor().toLowerCase() + "%"));
            }
            if (Objects.nonNull(bookFilter.getPrice())) {
                predicates.add(criteriaBuilder.equal(root.get("price"), bookFilter.getPrice()));
            }

            if (Objects.nonNull(bookFilter.getOrderBy()) && !bookFilter.getOrderBy().isBlank()) {
                query.orderBy(criteriaBuilder.desc(root.get(bookFilter.getOrderBy())));
            } else {
                query.orderBy(criteriaBuilder.asc(root.get("isIssued")));
            }
            return criteriaBuilder.and(predicates.toArray(new Predicate[0]));
        };
    }


}

File: src/main/kotlin/io/mcm/springbootwithkotlin/specification/BookSpecification.kt
package io.mcm.springbootwithkotlin.specification

import io.mcm.springbootwithkotlin.model.BooksRequest
import io.mcm.springbootwithkotlin.model.entities.Book
import org.springframework.data.jpa.domain.Specification
import org.springframework.stereotype.Component
import java.util.*
import javax.persistence.criteria.CriteriaBuilder
import javax.persistence.criteria.CriteriaQuery
import javax.persistence.criteria.Predicate
import javax.persistence.criteria.Root

@Component
class BookSpecification {
    fun build(request: BooksRequest): Specification<Book?> {
        return Specification { root: Root<Book?>, query: CriteriaQuery<*>, criteriaBuilder: CriteriaBuilder ->
            if (Objects.isNull(request) || Objects.isNull(request.bookFilter)) {
                println("The request book parameters are not provided")
                return@Specification criteriaBuilder.and(*arrayOfNulls(0))
            }
            val bookFilter = request.bookFilter
            val predicates: MutableList<Predicate> =
                ArrayList()
            if (Objects.nonNull(bookFilter!!.isIssued)) {
                predicates.add(criteriaBuilder.equal(root.get<Any>("isIssued"), bookFilter.isIssued))
            }
            if (Objects.nonNull(bookFilter.author) && !bookFilter.author!!.isBlank()) {
                predicates.add(
                    criteriaBuilder.like(
                        criteriaBuilder.lower(root.get("author")),
                        "%" + bookFilter.author!!.lowercase(Locale.getDefault()) + "%"
                    )
                )
            }
            if (Objects.nonNull(bookFilter.price)) {
                predicates.add(criteriaBuilder.equal(root.get<Any>("price"), bookFilter.price))
            }
            if (Objects.nonNull(bookFilter.orderBy) && !bookFilter.orderBy!!.isBlank()) {
                query.orderBy(criteriaBuilder.desc(root.get<Any>(bookFilter.orderBy)))
            } else {
                query.orderBy(criteriaBuilder.asc(root.get<Any>("isIssued")))
            }
            criteriaBuilder.and(*predicates.toTypedArray())
        }
    }
}